lib/maths/unittest/CBayesianOptimisationTest.cc

```diff
@@ -260,7 +260,7 @@ void CBayesianOptimisationTest::testMaximumExpectedImprovement() {
     }
 
     LOG_DEBUG(<< "mean gain = " << maths::CBasicStatistics::mean(gain));
-    CPPUNIT_ASSERT(maths::CBasicStatistics::mean(gain) > 1.29);
+    CPPUNIT_ASSERT(maths::CBasicStatistics::mean(gain) > 1.27);
```

Comment:
I did a bit of investigation into why this was necessary.  The trace logging from this test running with [Boost 1.65.1](https://github.com/elastic/ml-cpp/files/3573964/1.65.1.txt) and with [Boost 1.71](https://github.com/elastic/ml-cpp/files/3573967/1.71.txt) first shows a discrepancy on line 587 of those files.  For Boost 1.65.1:

```
CBayesianOptimisation.cc@104 x =     0.46252    0.105039   -0.372538 -0.00963219 EI(x) = -0.007086
```

and for Boost 1.71:

```
CBayesianOptimisation.cc@104 x =     0.46252    0.105039   -0.372538 -0.00963219 EI(x) = -0.0070849
```

The difference in `EI(x)` is tiny, but once the difference has occurred it propagates, eventually causing completely different code paths to be taken.

It implies there is a difference in the values being returned for the normal distribution PDF or CDF between the two Boost versions.  There has been no change to `boost/math/distributions/normal.hpp`.  However, that file in turn depends on `boost/math/special_functions/erf.hpp` and `boost/math/special_functions/gamma.hpp`, and there have been changes to those files:

* https://github.com/boostorg/math/commit/e348578b0f70f559ac3692f97aa05ba9b3235449
* https://github.com/boostorg/math/commit/08b7a6145f90eff611a1afbdd21e9c54faebcc27
* https://github.com/boostorg/math/commit/0152cfa2a9f9a586568e1194b125f85d9bc9d870

So it's most likely one of those that causes the difference.

@tveasey are you happy to proceed with the upgrade?

Reply:
Yes, I think this is fine. Qualitatively the difference in the test result is small. The difference is likely due to some improvement to numerics of of one of these functions.